### ISSUE TITLE
[Q] Can we access the hyperparameters used to run Sweeps?

### ISSUE THREAD

**Issue by naga-karthik** [wandb/wandb#4792]
Hello there, I was wondering if there's any way we can access the hyperparameters that `wandb agent` runs during Sweeps? I would like to create a python script that calls `wandb sweep`, runs `wandb agent`, gets a set of hyperparameters for that run, and uses those hyperparamters for running another downstream task. Is something like this possible with how Sweeps are defined currently?

I also looked into the high-level overview of sweeps in this [comment](https://github.com/wandb/wandb/issues/4586#issuecomment-1343170084) and it seems that in my plan above, I am currently bottlenecked by my understanding of how the Sweep Controller samples hyperparameters (that the agent subsequently runs). Any help on this is greatly appreciated, thank you in advance!

**Comment by thanos-wandb** [comment]
Hi @naga-karthik thanks a lot for your question! The `agent` isn't currently exposed through our API, therefore it won't be possible to get its attributes externally. It's only feasible if your downstream task is being called by the main function as in this [example](https://docs.wandb.ai/guides/sweeps/quickstart#set-up-your-training-code). This main function can access the `wandb.config` as there's a run initiated within it. Would this work for you? If not, could you please provide a bit more context on what this task is doing?

**Comment by naga-karthik** [comment]
Hey @thanos-wandb, thanks for getting back! Sure, here's some more context. Essentially, I am trying to integrate wandb's Sweeps functionality into a package that trains a model by specifying configuration `.json` files that contain all the necessary hyperparameters. What I had in mind was that if I could retrieve the sampled hyperparamters from wandb agent, then I can create a new configuration `.json` file for each combination of hyperparameters defined in `sweep_configuration` and then essentially launch multiple instances of my package.

Here's the idea in terms of pseudo-code:

```python

# define sweep configuration
sweep_configuration = {
    'method': 'random',
    'metric': {...},
    'parameters': {'batch_size': {'values': [16, 32, 64]}}
}

# get sweep id
sweep_id = wandb.sweep(sweep_configuration, project=project_name)

# run agent
wandb.agent(sweep_id, count=5)

# retrieve hyperparameters from the agent 
....

# create config file(s) based on retrieved hyperparamters
config_1 = ...

# launch my package with config_1
...

```
Please let me know if this makes sense! Hope I have not confused you even more!

I also have a follow-up question. When you say:
> This main function can access the wandb.config as there's a run initiated within it

Does this mean that the hyperparameters that wandb agent uses is stored in `wandb.config` for each combination of hyperparameters (which then becomes a run in Sweeps dashboard)? If yes, then if we have access to `wandb.config` (after running `wandb.agent`), we also can get the hyperparameters that will be used for sweeps right? Is my understanding correct?

**Comment by naga-karthik** [comment]
Hey @thanos-wandb, just wondering if you'd have some suggestions whether something could be done here? Thanks!

**Comment by scottire** [comment]
Hi @naga-karthik,
Have you tried getting the configs from `wandb.config`. 

# define sweep configuration
sweep_configuration = {
    'method': 'random',
    'metric': {...},
    'parameters': {'batch_size': {'values': [16, 32, 64]}}
}

# get sweep id
sweep_id = wandb.sweep(sweep_configuration, project=project_name) 

# run agent
wandb.agent(sweep_id, count=5)

# retrieve hyperparameters from the agent 
# create config file(s) based on retrieved hyperparamters
config_1 = wandb.config

# launch my package with config_1
...

**Comment by naga-karthik** [comment]
Hey @scottire, thanks for your suggestion and apologies for the delay in my response! I tried your suggestion using a toy example in wandb's docs and I have one question:

I noticed that after running `wandb.agent(sweep_id, count=5)`, the program first completes the 5 runs using the agent before getting to `config_1 = wandb.config`. This is rather unintended for my use case since I thought there could be a way to just get the combination of the hyperparamter values from `wand.agent`, and _then_ do the sweeps? I'm interested in the specific hyperparamter values that are used instead of the final result (of using some combination) on the Sweeps dashboard.

**Comment by thanos-wandb** [comment]
Hi @naga-karthik thanks for the additional context and the pseudo-code, that's really helpful. I wanted to follow up here and provide you with the following ways to access the values, each solution depends on the phase you would want to access the hyperparameters:

1) after `wandb.agent` completed, that's feasible using the API:

```
api = wandb.Api()

sweep = api.sweep("thanos-wandb/my-first-sweep/3o9w3c3q")
sweep_runs = sweep.runs

for sr in sweep_runs:
  print(sr.config)
```


2) during `wandb.agent` [the agent calls a main function](https://docs.wandb.ai/guides/sweeps/quickstart) that will consecutively initiate the experiments. You could access for each run the `wandb.config` inside the main function, and adjust your code so as to append to a dictionary argument. 

3) before executing `wandb.agent`, you could mock the sweep runs with the `sweeps` [package](https://github.com/wandb/sweeps/), which you can install as `pip install sweeps`. A minimal code example to get the runs, and their config: 

```
from sweeps import next_run, stop_runs, SweepRun, next_runs

sweep_config = {
        'method': 'grid',
        'parameters': {'a': {'values': [1, 2, 3]}}
}

suggested_run = next_runs(sweep_config, [], n = 3)
display(suggested_run)

for sr in suggested_run:
  print(sr.config)
```

Please let us know which option would best work for you, and if you have any further questions.

**Comment by naga-karthik** [comment]
Hey @thanos-wandb, thank you very much for getting back! Your suggestions are really helpful and option 3 is just what I was looking for! 

Just a quick follow up question - let's say I go ahead and get the mock configs from the sweeps package and run my downstream task (which only uses `wandb.log` for tracking my model training) using these configs. Then, unless I am running, `wandb.agent` I won't be seeing my results on the Sweeps dashboard right? 

In other words, does the sweeps package have any link to the Sweeps dashboard such that I don't have use `wandb.agent`? The thing is, the sweeps dashboard has useful information about the hyperparam combinations so I would like to get that, but, just wondering if there is a way to bypass `wandb.agent`? Thanks again in advance for your help!

**Comment by thanos-wandb** [comment]
Hi @naga-karthik glad you've found this useful, and that it works for you! The `wandb.agent` helps to orchestrate all the runs for the certain sweep, and will automatically create a parallel coordinates plot in your project page. However, if you wanted to avoid using `wandb.agent` and you are running individually `wandb.run` from your downstream task, you should be still able to `Add panel` > `Parallel coordinates plot` from the UI (project level). You can manually then add the config parameters to this chart, would this work for you?

**Comment by thanos-wandb** [comment]
Hi @naga-karthik just checking in here to see if the above would work for you, and if you have any further questions that we could help with? thanks!

**Comment by thanos-wandb** [comment]
Hi @naga-karthik since we haven't heard back from you in a while, I will go ahead and close this ticket for now. If you still have any questions, please let us know and we will be happy to keep investigating!
